include/ofi_mr.h

```diff
@@ -0,0 +1,118 @@
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <inttypes.h>
+
+#include "fi.h"
+#include "fi_util.h"
+
+/*
+ * MR monitor
+ */
+
+struct ofi_mr_subscription;
+struct ofi_mr_monitor_eq;
+
+enum ofi_mr_monitor_event {
+	OFI_MR_MONITOR_EVENT_MAPPED	= 1 << 1,
+	OFI_MR_MONITOR_EVENT_UNMAPPED	= 1 << 2,
+};
+
+struct ofi_mr_monitor {
+	void			*notifier;
+	fastlock_t		lock;
+	struct dlist_entry	eq_list;
+
+	int (*subscribe)(struct ofi_mr_monitor *mr_monitor, void *addr,
+			 uint64_t len, struct ofi_mr_subscription *subscribe);
+	void (*unsubscribe)(struct ofi_mr_monitor *mr_monitor,
+			    struct ofi_mr_subscription *subscribe);
+	struct ofi_mr_subscription *(*get_event)(struct ofi_mr_monitor_eq *eq);
+};
+
+struct ofi_mr_monitor_eq {
+	struct ofi_mr_monitor		*mr_monitor;
+	struct dlist_entry		entry;
+	int				events;
+	fastlock_t			lock;
+	struct dlist_entry		list;
+};
+
+struct ofi_mr_subscription {
+	struct ofi_mr_monitor_eq	*eq;
+	enum ofi_mr_monitor_event	event;
+	/* this entry is used either for
+	 * eq::subscirbe_list or for eq::read_list */
+	struct dlist_entry		entry;
+};
+
+int ofi_mr_monitor_init(struct ofi_mr_monitor *mr_monitor);
+void ofi_mr_monitor_finalize(struct ofi_mr_monitor *mr_monitor);
+void ofi_mr_monitor_eq_add(struct ofi_mr_monitor *mr_monitor,
+			   struct ofi_mr_monitor_eq *eq,
+			   int events);
+void ofi_mr_monitor_eq_del(struct ofi_mr_monitor_eq *eq);
+struct ofi_mr_subscription *ofi_mr_monitor_get_event(struct ofi_mr_monitor_eq *eq);
+
+static inline
+int ofi_mr_monitor_subscribe(struct ofi_mr_monitor_eq *eq,
+			     void *addr, uint64_t len,
+			     struct ofi_mr_subscription *subscript)
+{
+	int ret;
+
+	ret = eq->mr_monitor->subscribe(eq->mr_monitor, addr, len, subscript);
+	if (OFI_LIKELY(!ret))
+		FI_DBG(&core_prov, FI_LOG_MR,
+		       "monitoring addr=%p len=%"PRIu64" subscript=%p eq=%p\n",
+		       addr, len, subscript, eq);
+	else
+		FI_WARN(&core_prov, FI_LOG_MR,
+			"Failed (ret = %d) to monitor addr=%p len=%"PRIu64" subscript=%p eq=%p\n",
+			ret, addr, len, subscript, eq);
+	return ret;
+}
+
+static inline
+void ofi_mr_monitor_unsubscribe(struct ofi_mr_subscription *subscript)
+{
+	subscript->eq->mr_monitor->unsubscribe(subscript->eq->mr_monitor,
```

Comment:
What happens if the unsubscribe fails? Can the unsubscribe fail?

Reply:
Unsubcribe cannot fail.  There's nothing a user can use do with a failure code here.

Reply:
> Unsubcribe cannot fail.

If the underlying implementation is queue based, and completes requests asynchronously, then it is possible the request queue could fill and the unsubscribe request could fail. It's unlikely, but possible. 

If the assumption is that the unsubscribe function itself is responsible for handling failures, then it should be documented in the header file as a comment. 

Is it safe to assume that providers are responsible for providing the functions for the monitor? If yes, that also needs to be called out in the header file/documentation along with other expectations for provider writers. 

> There's nothing a user can use do with a failure code here.

A warning/info/debug should probably be printed in the event that something did not occur as expected. If not handled here, it would probably need to be handled within the function pointer provided to the monitor.

Reply:
The underlying implementation must do everything in its power to avoid unsubscribe failures.  If this means allocating something during subscribe to avoid or handle potential failures, then it needs to do that.  Any warnings should occur in the underlying monitor, which can also provide the real reason for any error.

The monitor is not associated with a provider, but the libfabric core.  The common initialization code will need to invoke the monitor specific initialization/cleanup routines.  A provider can select which monitor to use (if we ever allow enabling more than one at a time), but the monitor should be considered an internal 'service' available to all providers.